src/libsyntax_ext/test_harness.rs

```diff
@@ -77,49 +71,46 @@ impl<'a> MutVisitor for TestHarnessGenerator<'a> {
     }
 
     fn flat_map_item(&mut self, i: P<ast::Item>) -> SmallVec<[P<ast::Item>; 1]> {
-        let ident = i.ident;
-        if ident.name != kw::Invalid {
-            self.cx.path.push(ident);
-        }
-        debug!("current path: {}", path_name_i(&self.cx.path));
-
         let mut item = i.into_inner();
         if is_test_case(&item) {
             debug!("this is a test item");
 
             let test = Test {
                 span: item.span,
-                path: self.cx.path.clone(),
+                ident: item.ident,
             };
-            self.cx.test_cases.push(test);
-            self.tests.push(item.ident);
+            self.tests.push(test);
         }
 
         // We don't want to recurse into anything other than mods, since
         // mods or tests inside of functions will break things
         if let ast::ItemKind::Mod(mut module) = item.node {
             let tests = mem::take(&mut self.tests);
-            let tested_submods = mem::take(&mut self.tested_submods);
             noop_visit_mod(&mut module, self);
-            let tests = mem::replace(&mut self.tests, tests);
-            let tested_submods = mem::replace(&mut self.tested_submods, tested_submods);
+            let mut tests = mem::replace(&mut self.tests, tests);
 
-            if !tests.is_empty() || !tested_submods.is_empty() {
-                let (it, sym) = mk_reexport_mod(&mut self.cx, item.id, tests, tested_submods);
-                module.items.push(it);
-
-                if !self.cx.path.is_empty() {
-                    self.tested_submods.push((self.cx.path[self.cx.path.len()-1], sym));
+            if !tests.is_empty() {
+                let parent = if item.id == ast::DUMMY_NODE_ID {
+                    ast::CRATE_NODE_ID
                 } else {
-                    debug!("pushing nothing, sym: {:?}", sym);
-                    self.cx.toplevel_reexport = Some(sym);
+                    item.id
+                };
+                // Create an identifier that will hygienically resolve the test
+                // case name, even in another module.
+                let sp = self.cx.ext_cx.resolver.span_for_ast_pass(
+                    module.inner,
+                    AstPass::TestHarness,
+                    &[],
+                    Some(parent),
+                );
+                let expn = sp.ctxt().outer_expn();
+                for test in &mut tests {
+                    test.ident.span = test.ident.span.apply_mark(expn, Transparency::Opaque);
```

Comment:
I'd prefer to avoid using `apply_mark` directly and use the `with_def_site_ctxt`/`with_call_site_ctxt` functions instead.
They require setting the current expansion id, but that's maybe even for the better.
In general, the more this looks like a regular procedural macro, the better; perhaps we'll end up with all these passes actually being macros eventually (perhaps multiple of them per pass).

Reply:
Something like this https://github.com/petrochenkov/rust/commit/fb4172846d9098e6a5bf9f24687d643a4a29a0cd.

(Disregard the "setting the current expansion id" part, that's probably too annoying.)

Reply:
> I'd prefer to avoid using `apply_mark` directly and use the `with_def_site_ctxt`/`with_call_site_ctxt` functions instead.

I've added a comment about this. This uses `apply_mark` because it has to keep the original `SyntaxContext`.